Added cookbook for ragas integration

AnushreeBannadabhavi opened the pull request:
A cookbook showcasing the Haystack-Ragas integration. Fixes the issue [Cookbook for Ragas](https://github.com/deepset-ai/haystack-cookbook/issues/52)
- fixes  https://github.com/deepset-ai/haystack-cookbook/issues/52

AnushreeBannadabhavi commented:
Hi @TuanaCelik! Would love your feedback on my PR. It fixes https://github.com/deepset-ai/haystack-cookbook/issues/52.

vblagoje commented on notebooks/rag_eval_ragas.ipynb:
```diff
@@ -0,0 +1,5272 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    
</pre><p>Missing os import here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

vblagoje replied:
<pre class="ql-syntax" spellcheck="false">Line #2.    
</pre><p>nitpick: Let's do pip install as other notebooks, one liner with ! i.e.</p><p>!pip install "pydantic&lt;1.10.10", datasets&gt;=2.6.1", haystack-ai, ragas-haystack</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

vblagoje replied:
<p>Please add a short note here that we are using SQUAD here for convenience and tutorial purposes and that users will likely eval their domain specific datasets. </p><p>Also note what are the annotation requirements needed in a dataset, i.e which datasets fields are what you described further below with "Now that we have the <code>questions</code>, <code>generated answers</code>, <code>contexts</code> and the <code>ground truths</code>, we can begin our pipeline evaluation and compute all the supported metrics."</p><p> </p><p>It is prudent to notify users of these requirements before they proceed further. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

vblagoje replied:
<pre class="ql-syntax" spellcheck="false">Line #24.    os.environ["OPENAI_API_KEY"] = getpass("Enter OpenAI API key:")
</pre><p>No need to do this again, you've already completed intake of the OPENAI_API_KEY above. So either here or there, either is fine, but not both.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

vblagoje replied:
<p>Let's have a nice metric definition like you do in other sections :-) </p><p>Context Utilization is.... </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

vblagoje replied:
<p>Here as well, let's make definitions consistent</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

vblagoje replied:
<p>nitpick: definition</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

vblagoje replied:
<p>nitpick definition</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

vblagoje replied:
<p>nitpick: definition</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

vblagoje commented on notebooks/rag_eval_ragas.ipynb:
```diff
@@ -0,0 +1,5272 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #7.    
</pre><p>I also got deprecation warnings when I ran this part:</p><p>WARNING:ragas.metrics._context_precision:Using 'context_precision' without ground truth will be soon depreciated. Use 'context_utilization' instead</p><p> </p><p> </p><p>Would you please check if these can be resolved?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

AnushreeBannadabhavi replied:
<p>I tried resolving this, but wasn't able to.</p>

AnushreeBannadabhavi replied:
Looks like this can't be resolved because it comes from the ragas code itself. Analysis below.

The warning comes from this [line](https://github.com/explodinggradients/ragas/blob/5050742fc8a5c4ba80e6f899dea87de1f64601c6/src/ragas/metrics/_context_precision.py#L98) in the ragas code. 

- There's not a lot inside `ContextUtilization` [class](https://github.com/explodinggradients/ragas/blob/5050742fc8a5c4ba80e6f899dea87de1f64601c6/src/ragas/metrics/_context_precision.py#L177). It just inherits from `ContextPrecision` class. 
- The functionalities of `ContextUtilization` and `ContextPrecision` are differentiated based on whether ground truth answers are provided.
- Inside `ContextPrecision`, [`_get_row_attributes()`](https://github.com/explodinggradients/ragas/blob/5050742fc8a5c4ba80e6f899dea87de1f64601c6/src/ragas/metrics/_context_precision.py#L94) returns
      -  either row["question"], row["contexts"], row["**answer**"] for context utilization
      -  or returns row["question"], row["contexts"], row["**ground truth**"] for context precision
- Since no `ground truth answers` are provided for context utilization, the above warning is issued.

vblagoje commented on notebooks/rag_eval_ragas.ipynb:
```diff
@@ -0,0 +1,5272 @@
+{
```
<p>Why quotes around the name?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

AnushreeBannadabhavi replied:
<p>Removed</p>

vblagoje commented on notebooks/rag_eval_ragas.ipynb:
```diff
@@ -0,0 +1,5272 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #8.            "strictness": 2
</pre><p>What's strictness?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

AnushreeBannadabhavi replied:
<p>Updated the metric description with the definition</p>

vblagoje commented on notebooks/rag_eval_ragas.ipynb:
```diff
@@ -0,0 +1,5272 @@
+{
```
<p>Why not do all at once - for example all metrics that don't require ground truth or need common fields?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

AnushreeBannadabhavi replied:
<p>I'm not entirely clear on your question. Could you provide more details?</p><p>I included this section to demonstrate the computation of multiple metrics using a single pipeline object.</p>

vblagoje reviewed (changes requested):
@AnushreeBannadabhavi thanks a lot for this effort and apologies for delayed review. Although there are some minor request for changes this is already a solid work and an amazing contribution. Let's polish up some of the details and integrate this one 🚀

vblagoje commented:
@AnushreeBannadabhavi as we are, unfortunately, time bound for the integration of this PR would you please make these updates until tomorrow (12.4.24)? If not, I'll make them myself and we'll integrate this PR. Let us know.

AnushreeBannadabhavi commented:
Hi @vblagoje! Yes, I'll make the updates by tomorrow

vblagoje commented:
Awesome, thank you @AnushreeBannadabhavi

AnushreeBannadabhavi commented:
Hi @vblagoje! I've addressed all your review comments except the last one (multiple evaluator components). I can take a look at that quickly once I hear back from you.

vblagoje commented:
Thanks a lot @AnushreeBannadabhavi for a quick turnaround! @TuanaCelik the notebook, looks ok to me. I'll let you have the final word and integrate

vblagoje commented:
@davidsbatista this one is very similar to DeepEval, would you please also have a look?

davidsbatista commented:
@vblagoje LGTM, approved.

I was confused because all the "changes" in the review were still "open" - I think it's better to hit the button "mark as solved" if it's already solved.
